File: cpp/device-handling/device-list-changed/src/device-list-changed.cpp

#include <iostream>
#include <cstdlib>
#include <string>
#include <cstdio>>

#include <ic4/ic4.h>

void device_list_changed_handler(ic4::DeviceEnum& enumerator)
{
	auto device_list = ic4::DeviceEnum::getDevices();

	auto new_device_count = device_list.size();

	std::cout << "Device list has changed!" << std::endl;
	std::cout << "Found " << new_device_count << " devices" << std::endl;
	std::cout << std::endl;
}

int main()
{
	ic4::InitLibrary();
	std::atexit(ic4::ExitLibrary);

	ic4::DeviceEnum enumerator;
	auto token = enumerator.eventAddDeviceListChanged(device_list_changed_handler);

	auto initial_device_count = ic4::DeviceEnum::getDevices().size();

	std::cout << "Press ENTER to exit program" << std::endl;
	std::cout << initial_device_count << " devices connected initially." << std::endl;
	std::cout << std::endl;

	(void)std::getchar();

	// This is technically not necessary, since the DeviceEnum object is destroyed at the end of this function.
	enumerator.eventRemoveDeviceListChanged(token);

	return 0;
}

File: python/device-handling/device-list-changed/device-list-changed.py
import ic4


def handle_device_list_changed(device_enum: ic4.DeviceEnum):
    print("Device list changed!")

    print(f"Found {len(ic4.DeviceEnum.devices())} devices")

    print(ic4.DeviceEnum.devices())


def example_device_list_changed():
    enumerator = ic4.DeviceEnum()

    token = enumerator.event_add_device_list_changed(handle_device_list_changed)

    print("Waiting for DeviceListChanged event")
    print("Press Enter to exit")
    input()

    # Technically, this is not necessary, because the enumerator object is deleted when the function is exited
    # But for demonstration purposes, the event handler is removed:
    enumerator.event_remove_device_list_changed(token)


if __name__ == "__main__":
    ic4.Library.init(api_log_level=ic4.LogLevel.INFO, log_targets=ic4.LogTarget.STDERR)

    try:
        example_device_list_changed()
    finally:
        ic4.Library.exit()
